File: app/src/main/java/com/unsupportedpastels/hermesandroid/ui/BackgroundTaskStrip.kt
package com.unsupportedpastels.hermesandroid.ui

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.unsupportedpastels.hermesandroid.gateway.BackgroundTasks
import kotlinx.coroutines.delay

/** Child lifecycle is not the parent response's tool timeline. No speculative controls. */
@Composable
internal fun BackgroundTaskStrip(tasks: BackgroundTasks, nowOverride: Long? = null) {
    var dismissed by rememberSaveable { mutableStateOf(emptyList<String>()) }
    var clockNow by remember { mutableLongStateOf(System.currentTimeMillis()) }
    val now = nowOverride ?: clockNow
    LaunchedEffect(nowOverride, tasks.rows.isNotEmpty()) {
        if (nowOverride == null && tasks.rows.isNotEmpty()) while (true) {
            clockNow = System.currentTimeMillis()
            delay(1_000)
        }
    }
    // A dismissed row is hidden only while the exact observed evidence remains
    // dismissible. New evidence for the same child gets a different key and is
    // visible again; a row that becomes fresh/available is never suppressed.
    val visible = tasks.rows.filterNot { row ->
        row.isDismissible(now) && row.dismissalKey() in dismissed
    }
    if (visible.isEmpty()) return
    val presentation = tasks.presentation(visible, now)
    val completedOnly = presentation.terminalOnly
    val secondaryLabel = tasks.secondaryLabel(visible, now)
    var expanded by rememberSaveable(completedOnly) { mutableStateOf(false) }
    Surface(
        color = MaterialTheme.colorScheme.surfaceContainerHigh,
        shape = MaterialTheme.shapes.medium,
        modifier = Modifier.fillMaxWidth().testTag("Background task strip"),
    ) {
        Column(Modifier.padding(horizontal = 12.dp, vertical = 4.dp)) {
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                Column(Modifier.weight(1f).padding(top = 8.dp)) {
                    Text(presentation.headline, style = MaterialTheme.typography.labelLarge)
                    Text(
                        secondaryLabel,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                TextButton(onClick = { expanded = !expanded }) { Text(if (expanded) "Hide details" else "Details") }
            }
            if (expanded) {
                Column(Modifier.fillMaxWidth().heightIn(max = 176.dp).verticalScroll(rememberScrollState()),
                    verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    visible.forEach { row ->
                        Column {
                            Text(row.goal, style = MaterialTheme.typography.labelLarge)
                            Text(row.label(now), style = MaterialTheme.typography.bodySmall)
                            Text(
                                row.timeLabel(now),
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                            row.action?.let { Text(it, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant) }
                        }
                    }
                    Text("Only observed child events are shown. Silence does not mean finished or needs input.",
                        style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    val unavailable = visible.filter { !it.terminal && it.isDismissible(now) }
                    if (unavailable.isNotEmpty()) TextButton(onClick = {
                        dismissed = (dismissed + unavailable.map { it.dismissalKey() }).takeLast(64)
                    }) { Text("Dismiss unavailable") }
                    val completed = visible.filter { it.terminal }
                    if (completed.isNotEmpty()) TextButton(onClick = {
                        dismissed = (dismissed + completed.map { it.dismissalKey() }).takeLast(64)
                    }) { Text("Dismiss completed") }
                }
            }
        }
    }
}

File: ios/Mercury/Views/BackgroundTaskStrip.swift
import SwiftUI

struct BackgroundTaskStrip: View {
    let tasks: BackgroundTasks
    var nowOverride: Int64? = nil
    @State private var expanded = false
    @State private var dismissed: Set<String> = []

    var body: some View {
        TimelineView(.periodic(from: .now, by: 1)) { context in
            let now = nowOverride ?? Int64(context.date.timeIntervalSince1970 * 1000)
            // A dismissal hides only the exact observed evidence while it is
            // still unavailable. A later event, runtime rebind, or fresh row
            // has a different evidence key and is shown again.
            let visible = tasks.rows.filter {
                !($0.isDismissible(now: now) && dismissed.contains($0.dismissalKey))
            }
            if !visible.isEmpty {
                let presentation = tasks.presentation(rows: visible, now: now)
                let secondaryLabel = tasks.secondaryLabel(rows: visible, now: now)
                VStack(alignment: .leading, spacing: 8) {
                    HStack {
                        VStack(alignment: .leading, spacing: 3) {
                            Text(presentation.headline)
                                .font(.subheadline.weight(.semibold))
                                .lineLimit(2)
                            Text(secondaryLabel)
                                .font(.caption)
                                .foregroundStyle(.secondary)
                                .lineLimit(2)
                        }
                        Spacer(minLength: 4)
                        Button(expanded ? "Hide details" : "Details") { expanded.toggle() }
                            .font(.subheadline).frame(minHeight: 44)
                    }
                    if expanded {
                        ScrollView {
                            VStack(alignment: .leading, spacing: 10) {
                                ForEach(visible) { row in
                                    VStack(alignment: .leading, spacing: 3) {
                                        Text(row.goal).font(.subheadline.weight(.semibold))
                                        Text(row.label(now: now)).font(.caption)
                                        Text(row.timeLabel(now: now))
                                            .font(.caption).foregroundStyle(.secondary)
                                        if let action = row.action {
                                            Text(action).font(.caption).foregroundStyle(.secondary)
                                        }
                                    }.frame(maxWidth: .infinity, alignment: .leading)
                                }
                                Text("Only observed child events are shown. Silence does not mean finished or needs input.")
                                    .font(.caption).foregroundStyle(.secondary)
                                let unavailable = visible.filter {
                                    !$0.terminal && $0.isDismissible(now: now)
                                }
                                if !unavailable.isEmpty {
                                    Button("Dismiss unavailable") {
                                        dismissed.formUnion(unavailable.map(\.dismissalKey))
                                    }
                                    .frame(minHeight: 44)
                                }
                                let completed = visible.filter(\.terminal)
                                if !completed.isEmpty {
                                    Button("Dismiss completed") {
                                        dismissed.formUnion(completed.map(\.dismissalKey))
                                    }
                                    .frame(minHeight: 44)
                                }
                            }
                        }.frame(maxHeight: 176)
                    }
                }
                .padding(.horizontal, 12).padding(.vertical, 5)
                .background(.regularMaterial, in: RoundedRectangle(cornerRadius: 14))
                .accessibilityIdentifier("background-task-strip")
            }
        }
    }
}
